File: security/rbac.go
package security

import (
	"context"
	"net/http"

	"doctor-service/entities"
)

type contextKey string

const AdminRoleContextKey contextKey = "admin_role"
const AdminIDContextKey contextKey = "admin_id"

func RequireRole(allowedRoles ...entities.AdminRole) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			roleVal := r.Context().Value(AdminRoleContextKey)
			if roleVal == nil {
				// Check header fallback for local dev
				roleHeader := r.Header.Get("X-Admin-Role")
				if roleHeader != "" {
					roleVal = entities.AdminRole(roleHeader)
				}
			}

			if roleVal == nil {
				http.Error(w, `{"error":"Unauthorized: Admin authentication token required"}`, http.StatusUnauthorized)
				return
			}

			userRole := roleVal.(entities.AdminRole)
			authorized := false

			for _, role := range allowedRoles {
				if userRole == role || userRole == entities.AdminRoleSuperAdmin {
					authorized = true
					break
				}
			}

			if !authorized {
				http.Error(w, `{"error":"Forbidden: Insufficient administrative privileges to perform this action"}`, http.StatusForbidden)
				return
			}

			next.ServeHTTP(w, r)
		})
	}
}

func WithAdminContext(ctx context.Context, adminID string, role entities.AdminRole) context.Context {
	ctx = context.WithValue(ctx, AdminIDContextKey, adminID)
	return context.WithValue(ctx, AdminRoleContextKey, role)
}

File: python_backend/security/rbac.py
def require_role(*allowed_roles):
    def middleware(request, next_handler):
        role_val = request.context.get("admin_role")
        if not role_val:
            role_header = request.headers.get("X-Admin-Role")
            if role_header:
                role_val = role_header
                
        if not role_val:
            return {"error": "Unauthorized: Admin authentication token required"}, 401
            
        authorized = False
        for role in allowed_roles:
            if role_val == role or role_val == "SUPER_ADMIN":
                authorized = True
                break
                
        if not authorized:
            return {"error": "Forbidden: Insufficient administrative privileges to perform this action"}, 403
            
        return next_handler(request)
    return middleware

def with_admin_context(context, admin_id, role):
    context["admin_id"] = admin_id
    context["admin_role"] = role
    return context
